dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,194 @@
+OutType to_bit_mask(const BitIndices &bit_indices) {
+  OutType mask = 0;
+  for (int idx : bit_indices)
+    mask |= OutType(1) << idx;
+
+  return mask;
+}
+
+/**
+ * @param out_axes a collection of integers, supporting `clear` and `push_back`.
+ * @param dim_groups a collection of integer pairs, supporting `clear` and `push_back`
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
+                       const TensorListShape<ndim> &in_shape, span<const int> axes) {
+  out_axes.clear();
+  dim_groups.clear();
+  int d = in_shape.sample_dim();
+
+  uint64_t mask = to_bit_mask(axes);
+
+  int i;
+
+  // skip leading degenerate dimensions
+  for (i = 0; i < d && is_degenerate_dim(in_shape, i); i++) {}
+
+  if (i < d) {
+    // so we've reached the outermost non-degenerate dimension
+    int remapped = 0;  // its index will map to 0
+
+    bool prev_reduced = (mask >> i) & 1;  // is it reduced?
+    if (prev_reduced)
+      out_axes.push_back(remapped);  // if so, add the axis to the list of reduced dimension groups
+
+    int group_start = 0;     // the dimension group starts at 0 -
+    int group_size = i + 1;  // all leading degenerate dimensions are in this group
+
+    for (i++; i < d; i++) {
+      if (is_degenerate_dim(in_shape, i)) {
+        // Degenerate dimension can be merged with the previous one regardless of whether
+        // it is reduced or not.
+        group_size++;
+        continue;
+      }
+      bool is_reduced = (mask >> i) & 1;
+      if (is_reduced != prev_reduced) {
+        dim_groups.emplace_back(group_start, group_size);
+        group_start = i;
+        group_size = 1;
+        remapped++;
+        if (is_reduced)
+          out_axes.push_back(remapped);
+      } else {
+        group_size++;
+      }
+      prev_reduced = is_reduced;
+    }
+    dim_groups.push_back(std::make_pair(group_start, group_size));
+  } else {
+    // totally degenerate
+    dim_groups.push_back(std::make_pair(0, d));
+  }
+}
+
+
+inline void CheckBatchReduce(const TensorListShape<> &tls, span<const int> axes) {
+  if (tls.num_samples() == 0)
+    return;
+
+  uint64_t mask = to_bit_mask(axes);
+  SmallVector<int, 6> non_reduced;
+  for (int a = 0; a < tls.sample_dim(); a++) {
+    if (!(mask & (1 << a)))
+      non_reduced.push_back(a);
+  }
+
+  auto first_sample_shape = tls.tensor_shape_span(0);
+
+  for (int i = 1; i < tls.num_samples(); i++) {
+    auto sample_shape = tls.tensor_shape_span(i);
+    for (int a : non_reduced) {
+      DALI_ENFORCE(sample_shape[a] == first_sample_shape[a], make_string(
+        "Reduce: batch reduction requires that all samples have the same extent in non-reduced "
+        "dimensions.\nError at sample ", i, " axis ", a, ": the extent is ", sample_shape[a],
+        " != ", first_sample_shape[a], " in the first sample in the batch."));
+    }
+  }
+}
+
+
+inline void CheckAxes(span<const int> axes, int dim) {
```

Comment:
```suggestion
inline void CheckAxes(span<const int> axes, int ndims) {
```
`dim` IMHO suggests, it's an index of some specific dimension